nixos/modules/services/web-servers/nginx/default.nix

```diff
@@ -372,10 +372,10 @@ in
       preStart =  mkOption {
         type = types.lines;
         default = ''
-          test -d ${cfg.stateDir}/logs || mkdir -m 750 -p ${cfg.stateDir}/logs  
-          test `stat -c %a ${cfg.stateDir}` = "750" || chmod 750 ${cfg.stateDir}
-          test `stat -c %a ${cfg.stateDir}/logs` = "750" || chmod 750 ${cfg.stateDir}/logs
-          chown -R ${cfg.user}:${cfg.group} ${cfg.stateDir}
+          test -d '${cfg.stateDir}/logs' || mkdir -p -m 750 '${cfg.stateDir}/logs'
+          test `stat -c %a '${cfg.stateDir}'` = "750" || chmod 750 '${cfg.stateDir}'
+          test `stat -c %a '${cfg.stateDir}/logs'` = "750" || chmod 750 '${cfg.stateDir}/logs'
```

Comment:
wonder why `test` is required here. `chmod` and `mkdir -p` are idempotent, no?

Reply:
Normally these commands will be ran, when directories already exist. And "test" is faster than "mkdir", because it is built-in command in bash. So "test" makes service startup a bit faster, except for the first startup, when directories do not exist yet.

Reply:
This "a bit" is too a bit.

Also, this duplicates `tmpfiles` configuration. Why both?
Except for `logs`... But I don't see `logs` usage in this module. Looks like by default nginx logs errors to stderr (journald) and access log isn't set, so why create logs dir explicitly?

Reply:
Start nginx without logs folder - 
```
sudo -u nginx nginx -c /nix/store/hgfqd1mr72l3jhwv9kc76ls3jk5plqmh-nginx.conf -p /var/spool/nginx                                                                                                                                                                                  
nginx: [alert] could not open error log file: open() "/var/spool/nginx/logs/error.log" failed (2: No such file or directory)
2019/03/09 20:45:37 [emerg] 23512#23512: open() "/var/spool/nginx/logs/access.log" failed (2: No such file or directory)
```

Reply:
nginx requires a folder logs

Reply:
Try starting with the proper configuration file. By default, it redirects the log to stderr.

Reply:
See also: https://github.com/NixOS/nixpkgs/blob/532a80b94225065858855d75ba320c311b094b41/nixos/modules/services/web-servers/nginx/default.nix#L357

Reply:
Running with this config:
```
error_log stderr;
pid /run/nginx/nginx.pid;
daemon off;
events {
}
http {
  ssl_protocols TLSv1.2 TLSv1.3;
  ssl_ciphers EECDH+aRSA+AESGCM:EDH+aRSA:EECDH+aRSA:+AES256:+AES128:+SHA1:!CAMELLIA:!SEED:!3DES:!DES:!RC4:!eNULL;
  client_max_body_size 10m;
}
```
error:
```
ginx: [alert] could not open error log file: open() "/var/spool/nginx/logs/error.log" failed (2: No such file or directory)
2019/03/09 21:07:33 [emerg] 23844#23844: open() "/run/nginx/nginx.pid" failed (2: No such file or directory)
```

Reply:
@Izorkin indeed! Maybe then add it to tmpfiles config and remove here?

Reply:
@dasJ it sounds like a bug in nginx - it requires that logs directory even for `error_log stderr`

Reply:
Updated config.

Reply:
@Izorkin Looks great now! :shipit:

Reply:
Hold up, where did the config check go? Was this an accident or on purpose?